lib/request-proxy-send.js

```diff
@@ -0,0 +1,279 @@
+    }));
+};
+
+RequestProxySend.prototype.handleSuccess = function handleSuccess(res1, res2, callback) {
+    if (this.numRetries > 0) {
+        this.ringpop.stat('increment', 'requestProxy.retry.succeeded');
+        this.ringpop.logger.info('ringpop request proxy retry succeeded', {
+            local: this.ringpop.whoami(),
+            errors: this.errors,
+            numRetries: this.numRetries,
+            maxRetries: this.maxRetries,
+            reqMethod: this.reqMethod,
+            reqUrl: this.reqUrl,
+            reqStartTime: this.reqStartTime,
+            reqTotalTime: Date.now() - this.reqStartTime,
+            retrySchedule: this.retrySchedule,
+            retryStartTime: this.retryStartTime,
+            retryTotalTime: Date.now() - this.retryStartTime,
+            destinations: this.destinations
+        });
+
+        this.ringpop.emit('requestProxy.retrySucceeded');
+    }
+
+    callback(null, res1, res2);
+};
+
+RequestProxySend.prototype.lookupKeys = function lookupKeys(keys) {
+    var self = this;
+
+    var destEntries = keys.reduce(function reduceKey(dests, key) {
+        var dest = self.ringpop.lookup(key);
+        dests[dest] = true;
+        return dests;
+    }, {});
+
+    return Object.keys(destEntries);
+};
+
+RequestProxySend.prototype.rerouteRetry = function rerouteRetry(newDest, callback) {
+    this.destinations.push(newDest);
+
+    this.ringpop.logger.warn('ringpop request proxy retry rerouted', {
+        local: this.ringpop.whoami(),
+        errors: this.errors,
+        destOrig: this.channelOpts.host,
+        destNew: newDest,
+        destinations: this.destinations,
+        reqMethod: this.reqMethod,
+        reqUrl: this.reqUrl
+    });
+
+    this.ringpop.emit('requestProxy.retryRerouted', this.channelOpts.host, newDest);
+
+    if (newDest === this.ringpop.whoami()) {
+        this.ringpop.stat('increment', 'requestProxy.retry.reroute.local');
+        this.requestProxy.handleRequest(
+            this.getRawHead(),
+            this.request.body,
+            callback
+        );
+        return;
+    }
+
+    this.ringpop.stat('increment', 'requestProxy.retry.reroute.remote');
+
+    // NOTE Passes in different channel opts than all other calls to send
+    this.send({
+        host: newDest,
+        timeout: this.channelOpts.timeout,
+        endpoint: this.channelOpts.endpoint
+    }, callback);
+};
+
+RequestProxySend.prototype.scheduleRetry = function scheduleRetry(callback) {
+    var self = this;
+
+    if (this.numRetries === 0) {
+        this.retryStartTime = Date.now();
+    }
+
+    // `scheduleDelay` could result in NaN. If so, default to max.
+    var scheduleDelay = this.retries.schedule[this.numRetries] * 1000;
+    var delay = numOrDefault(scheduleDelay, this.maxRetryTimeout);
+
+    // TODO Can we make this delay responsive/adaptive to previously
+    // recorded cluster converge-times?
+    this.timeout = setTimeout(function onTimeout() {
```

Comment:
suggestion: To ease testing you might want to do `self.setTimeout()`

You can then use `time-mock` ( https://github.com/uber/tchannel/blob/master/node/test/timeouts.js#L23 ) to do integration tests without having actual timeouts. This will speed up your integration tests.